File: scala/octal/src/main/scala/Octal.scala
object Octal {

  def octalToInt(octal: String) = {
    require(octal.length > 0)
    require(octal.forall(_.asDigit < 8))

    octal.reverse.zipWithIndex.map { case (c, i) => octalDigit(c, i) }.sum
  }

  private def octalDigit(digit: Char, index: Int) = digit.asDigit * Math.pow(8, index)

  def intToOctal(decimal: Int) = octalDigits(decimal, Nil).mkString

  private def octalDigits(decimal: Int, digits: List[Int]): List[Int] =
    decimal / 8 match {
      case 0 => decimal :: digits
      case div => octalDigits(div, decimal % 8 :: digits)
    }
}

File: haskell/octal/Octal.hs
module Octal (showOct, readOct) where

import Data.Char (digitToInt)

showOct :: (Integral a, Show a) => a -> String
showOct n = showOctLoop n "" where
    showOctLoop num acc = acc `seq` case num `div` 8 of
        0 -> (show num) ++ acc
        x -> showOctLoop x $ (show (num `mod` 8) ++ acc)

readOct :: Integral a => String -> a
readOct str = str `seq` sum $ map readOctDigit $ zip (reverse str) [0..length str - 1] where
    readOctDigit (digit, index) = fromIntegral (digitToInt digit) * 8 ^ index